### ISSUE TITLE
link_error support for complex canvas

### ISSUE THREAD

**Issue by itayB** [celery/celery#4048]
## Checklist

- [√] I have included the output of ``celery -A proj report`` in the issue.
      (if you are not able to do this, then at least specify the Celery
       version affected).
- [ ] I have verified that the issue exists against the `master` branch of Celery.

## Setup
Celery 4.0.2, broker=RabbitMQ, backend=Redis
**UPDATE**: same for Celery 4.1

## Steps to reproduce
Try to add `link_error` in an apply_async of chain (with chord within). Till now I was using celery 3.1 and it didn't work at all. Now after upgrade to celery 4.0.2 I'm getting some unexpected behavior.
Having 13 different tasks working on a shared resource (products feed), I want to "lock" that resource on start and "unlock" it at the end. whenever task fails (and other concurrent tasks finish as well) I want to "unlock" the resource again.  

`celery worker -A worker.celeryapp:app -l info -Q default -c 2 -n defaultworker@%h -Ofair`

```
@app.task(name='task_1',
          bind=True,
          base=MyConnectionHolderTask)
def task_1(self, feed_id, flow_id, **kwargs):
    do_something()
    # raise Exception('test')
```

```
lock = lock_flow_task.si(feed_id, flow_id)
unlock = unlock_flow_task.si(feed_id, flow_id)
task_1 = t_1.si(feed_id, flow_id)
  .
  .
task_13 = t_13.si(feed_id, flow_id)


           (lock |
            task_1 |
            group((task_2 | task_3 | task_4),
                  task_5,
                  task_6,
                  task_7,
                  task_8) |
            task_9 |
            task_10 |
            task_11 |
            task_12 |
            task_13 |
            unlock).apply_async(link_error=unlock)
````

## Expected behavior
Each one of the tasks (task_1..task_13) failure should trigger finally the `unlock` task. If one of the concurrent `group` tasks is running - it should wait until all finished and run the `unlock` as well.

## Actual behavior
Trying to test the `link_error` for each task failure (by adding `raise Exception("test")` in each task.
Exception in `task_1` works fine -> `unlock` run.
Exception in `task_2` didn't work. Ends at the end of the grouped tasks.
Exception in `task_4` caused the `unlock` to run twice!
Exception in `task_5` caused the `unlock` to run twice!
Exception in `task_6` caused the `unlock` to run twice!
Exception in `task_9` works fine

Any idea or workaround to achieve the desired state?

log from the twice unlock:
```
2017-05-24 07:48:52,155 14     ERROR   Chord callback for '9d8005cb-8335-4738-b3ae-138717193e3d' raised: ChordError(u"Dependency a6c98e9f-87ac-40f2-9f91-51ad271d5a33 raised Exception(u'',)",) [file: redis.py, line: 284]
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/backends/redis.py", line 281, in on_chord_part_return
    callback.delay([unpack(tup, decode) for tup in resl])
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/backends/redis.py", line 234, in _unpack_chord_result
    raise ChordError('Dependency {0} raised {1!r}'.format(tid, retval))
ChordError: Dependency a6c98e9f-87ac-40f2-9f91-51ad271d5a33 raised Exception(u'test',)
2017-05-24 07:48:52,157 14     INFO    MyCeleryRouter: Inserted To Queue: lock, task: unlock_task, args:(727, 1495612112858) [file: celery_router.py, line: 64]
2017-05-24 07:48:52,165 14     ERROR   Chord '9d8005cb-8335-4738-b3ae-138717193e3d' raised: TypeError('sequence item 1: expected string, NoneType found',) [file: redis.py, line: 293]
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/backends/redis.py", line 287, in on_chord_part_return
    ChordError('Callback error: {0!r}'.format(exc)),
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/backends/base.py", line 223, in chord_error_from_stack
    return backend.fail_from_current_stack(callback.id, exc=exc)
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/backends/base.py", line 230, in fail_from_current_stack
    self.mark_as_failure(task_id, exc, ei.traceback)
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/backends/base.py", line 163, in mark_as_failure
    traceback=traceback, request=request)
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/backends/base.py", line 309, in store_result
    request=request, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/backends/base.py", line 652, in _store_result
    self.set(self.get_key_for_task(task_id), self.encode(meta))
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/backends/base.py", line 558, in get_key_for_task
    self.task_keyprefix, key_t(task_id), key_t(key),
TypeError: sequence item 1: expected string, NoneType found
2017-05-24 07:48:52,167 14     INFO    Inserted To Queue: lock, task: unlock_task, args:(727, 1495612112858) [file: celery_router.py, line: 64]
2017-05-24 07:48:52,167 1      INFO    Received task: unlock_task[74cf0ef9-4933-4ba4-b0fc-a9a4f50c3929]   [file: strategy.py, line: 109]
2017-05-24 07:48:52,171 1      INFO    Received task: unlock_task[2784cca9-1139-4690-9b8b-60fef2644881]   [file: strategy.py, line: 109]
2017-05-24 07:48:52,172 14     WARNING /usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/app/trace.py:542: RuntimeWarning: Exception raised outside body: TypeError('sequence item 1: expected string, NoneType found',):
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/app/trace.py", line 381, in trace_task
    I, R, state, retval = on_error(task_request, exc, uuid)
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/app/trace.py", line 323, in on_error
    task, request, eager=eager, call_errbacks=call_errbacks,
   File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/app/trace.py", line 157, in handle_error_state
    call_errbacks=call_errbacks)
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/app/trace.py", line 202, in handle_failure
    call_errbacks=call_errbacks,
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/backends/base.py", line 166, in mark_as_failure
    self.on_chord_part_return(request, state, exc)
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/backends/redis.py", line 296, in on_chord_part_return
    ChordError('Join error: {0!r}'.format(exc)),
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/backends/base.py", line 223, in chord_error_from_stack
    return backend.fail_from_current_stack(callback.id, exc=exc)
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/backends/base.py", line 230, in fail_from_current_stack
    self.mark_as_failure(task_id, exc, ei.traceback)
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/backends/base.py", line 163, in mark_as_failure
    traceback=traceback, request=request)
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/backends/base.py", line 309, in store_result
    request=request, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/backends/base.py", line 652, in _store_result
    self.set(self.get_key_for_task(task_id), self.encode(meta))
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/backends/base.py", line 558, in get_key_for_task
    self.task_keyprefix, key_t(task_id), key_t(key),
TypeError: sequence item 1: expected string, NoneType found

  exc, exc_info.traceback))) [file: log.py, line: 235]
2017-05-24 07:48:52,177 14     ERROR   Task SIMILARITY_TASK[a6c98e9f-87ac-40f2-9f91-51ad271d5a33] raised unexpected: TypeError('sequence item 1: expected string, NoneType found',) [file: trace.py, line: 241]
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/app/trace.py", line 381, in trace_task
    I, R, state, retval = on_error(task_request, exc, uuid)
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/app/trace.py", line 323, in on_error
    task, request, eager=eager, call_errbacks=call_errbacks,
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/app/trace.py", line 157, in handle_error_state
    call_errbacks=call_errbacks)
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/app/trace.py", line 202, in handle_failure
    call_errbacks=call_errbacks,
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/backends/base.py", line 166, in mark_as_failure
    self.on_chord_part_return(request, state, exc)
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/backends/redis.py", line 296, in on_chord_part_return
    ChordError('Join error: {0!r}'.format(exc)),
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/backends/base.py", line 223, in chord_error_from_stack
    return backend.fail_from_current_stack(callback.id, exc=exc)
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/backends/base.py", line 230, in fail_from_current_stack
    self.mark_as_failure(task_id, exc, ei.traceback)
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/backends/base.py", line 163, in mark_as_failure
    traceback=traceback, request=request)
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/backends/base.py", line 309, in store_result
    request=request, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/backends/base.py", line 652, in _store_result
    self.set(self.get_key_for_task(task_id), self.encode(meta))
  File "/usr/local/lib/python2.7/site-packages/celery-4.0.2-py2.7.egg/celery/backends/base.py", line 558, in get_key_for_task
    self.task_keyprefix, key_t(task_id), key_t(key),
TypeError: sequence item 1: expected string, NoneType found
2017-05-24 07:48:52,181 14     INFO    unlock_task    [2784cca9-1139-4690-9b8b-60fef2644881] unlock for feed 727, flow 1495612112858 [file: lock_flow_task.py, line: 50]
2017-05-24 07:48:52,183 14     INFO    Task unlock_task[2784cca9-1139-4690-9b8b-60fef2644881] succeeded in 0.00978600300732s: None [file: trace.py, line: 441]
2017-05-24 07:48:52,186 13     WARNING unlock_task    [74cf0ef9-4933-4ba4-b0fc-a9a4f50c3929] failed to unlock feed 727: expected flow 1495612112858, actual flow None [file: section_flow_lock.py, line: 67]
2017-05-24 07:48:52,188 13     INFO    Task unlock_task[74cf0ef9-4933-4ba4-b0fc-a9a4f50c3929] succeeded in 0.0141102350026s: None [file: trace.py, line: 441]
```

**Comment by nvictor** [comment]
Having a very similar issue. 

In our case, a group, linked to a post process chain on both success and error, runs in parallel. On error, the chain (of the group) that failed never synchronizes. All other chains synchronizes on success, making the post process chain being called multiple times.

**Comment by ocervell** [comment]
Same issue here.
This is making big workflows of tasks fail.

**Comment by xeor** [comment]
Not sure if exactly the same, but it for sure look extremely like my problem as well.

**Comment by auvipy** [comment]
anyone could verify the issue in recent master with latest dependencies?

**Comment by ocervell** [comment]
I haven't been able to pinpoint exactly under which conditions the workflows fail. 
@itayB, would you be able to test with the latest master with the workflow you described ?

**Comment by auvipy** [comment]
closing lack of feedback.

**Comment by jessecooper** [comment]
I ran into this issue on the latest master branch. I opened a new issue and have also included a hacky work around to fix the type exception. 
https://github.com/celery/celery/issues/5229
